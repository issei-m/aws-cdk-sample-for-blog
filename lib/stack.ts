import { Stack as BaseStack, type StackProps as BaseStackProps } from 'aws-cdk-lib'
import type { Construct } from 'constructs'

export interface StackProps extends BaseStackProps {
  resourceNamePrefix: string
}

export class Stack extends BaseStack {
  protected readonly prefix: string

  constructor (scope: Construct, id: string, props: StackProps) {
    super(scope, id, props)

    this.prefix = props.resourceNamePrefix
  }
}
