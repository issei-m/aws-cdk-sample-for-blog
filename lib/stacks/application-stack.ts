import * as cdk from 'aws-cdk-lib'
import { type Construct } from 'constructs'
import { StackProps, Stack } from '../stack'

export interface ApplicationStackProps extends StackProps {
  publicUserUploadedFilesBucket: cdk.aws_s3.IBucket
  functionMemorySize: number
}

export class ApplicationStack extends Stack {
  constructor (scope: Construct, id: string, props: ApplicationStackProps) {
    super(scope, id, props)

    const myHandlerFuncRole = new cdk.aws_iam.Role(this, 'ApplicationFunctionRole', {
      roleName: `${this.prefix}-application-function-role`,
      assumedBy: new cdk.aws_iam.ServicePrincipal('lambda.amazonaws.com')
    })
    myHandlerFuncRole.addManagedPolicy(cdk.aws_iam.ManagedPolicy.fromAwsManagedPolicyName('service-role/AWSLambdaBasicExecutionRole'))
    new cdk.aws_lambda.Function(this, 'ApplicationFunction', {
      role: myHandlerFuncRole,
      code: cdk.aws_lambda.Code.fromInline(`
import os

def handler(event, context):
    return { 'bucket': os.environ['BUCKET_NAME'] }
`.trim()),
      handler: 'index.handler',
      architecture: cdk.aws_lambda.Architecture.ARM_64,
      runtime: cdk.aws_lambda.Runtime.PYTHON_3_12,
      functionName: `${this.prefix}-application`,
      memorySize: props.functionMemorySize,
      environment: {
        BUCKET_NAME: props.publicUserUploadedFilesBucket.bucketName,
      }
    })
  }
}
