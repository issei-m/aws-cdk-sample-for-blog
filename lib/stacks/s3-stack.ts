import * as cdk from 'aws-cdk-lib'
import { RemovalPolicy } from 'aws-cdk-lib'
import { type Construct } from 'constructs'
import { Stack, type StackProps } from '../stack'

export class S3Stack extends Stack {
  public readonly publicUserUploadedFilesBucket: cdk.aws_s3.IBucket

  constructor (scope: Construct, id: string, props: StackProps) {
    super(scope, id, props)

    this.publicUserUploadedFilesBucket = new cdk.aws_s3.Bucket(this, 'PublicUserUploadedFilesBucket', {
      bucketName: `${this.prefix}-public-user-uploaded-files`,
      removalPolicy: RemovalPolicy.RETAIN
    })
  }
}
