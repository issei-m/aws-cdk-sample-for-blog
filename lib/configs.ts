import { type Stage, type Configuration } from './stage-context'

export const configs: Record<Stage, Configuration> = {
  prod: {
    applicationFunctionMemorySize: 1024,
  },
  dev: {
    applicationFunctionMemorySize: 128,
  }
}
