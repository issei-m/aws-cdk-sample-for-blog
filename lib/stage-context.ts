import { configs as availableConfigs } from './configs'
import { type IConstruct } from 'constructs/lib/construct'

export interface Configuration {
  applicationFunctionMemorySize: number
}

export type Stage = 'dev' | 'prod'

export class StageContext {
  public readonly stage: Stage
  public readonly config: Configuration
  private static instance: StageContext

  private constructor (stage: Stage, config: Configuration) {
    this.stage = stage
    this.config = config
  }

  public static getContext (construct: IConstruct): StageContext {
    if (StageContext.instance === undefined) {
      const stage = construct.node.tryGetContext('stage')
      if (stage === undefined) {
        throw new Error('The context [stage] must be provided, forgot to specify "--context stage=XXX" option?')
      }

      const config = availableConfigs[stage as Stage]
      if (config === undefined) {
        throw new Error(`The env "${stage as string}" is not available, it must be either of [${Object.keys(availableConfigs).join(', ')}]`)
      }

      StageContext.instance = new StageContext(stage as Stage, config)
    }

    return StageContext.instance
  }
}
