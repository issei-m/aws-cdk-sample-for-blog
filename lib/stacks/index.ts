import * as cdk from 'aws-cdk-lib'
import {Configuration, Stage} from '../stage-context'
import { S3Stack } from './s3-stack'
import { ApplicationStack } from './application-stack'

export interface Stacks {
  s3Stack: S3Stack
  applicationStack: ApplicationStack
}

export function synthesizeStacks (
  app: cdk.App,
  resourceNamePrefix: string,
  config: Configuration,
  buildStackName: (baseName: string) => string,
): Stacks {
  const baseStackProps = {
    resourceNamePrefix: resourceNamePrefix,
  }

  const s3Stack = new S3Stack(app, buildStackName('S3'), {
    ...baseStackProps,
  })

  const applicationStack = new ApplicationStack(app, buildStackName('Application'), {
    ...baseStackProps,
    publicUserUploadedFilesBucket: s3Stack.publicUserUploadedFilesBucket,
    functionMemorySize: config.applicationFunctionMemorySize,
  })

  return {
    s3Stack,
    applicationStack,
  }
}
