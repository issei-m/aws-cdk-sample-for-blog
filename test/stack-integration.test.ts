import * as cdk from 'aws-cdk-lib'
import { synthesizeStacks } from '../lib/stacks'
import { Template } from 'aws-cdk-lib/assertions'

describe('Stack integration testing', () => {
  const app = new cdk.App()

  const stacks = synthesizeStacks(
    app,
    'test',
    {
      applicationFunctionMemorySize: 1024,
    },
    (baseName: string) => `Test${baseName}`,
  )

  describe('S3Stack', () => {
    test('synthesizes the way we expect', () => {
      expect(Template.fromStack(stacks.s3Stack)).toMatchSnapshot()
    })
  })

  describe('ApplicationStack', () => {
    test('synthesizes the way we expect', () => {
      expect(Template.fromStack(stacks.applicationStack)).toMatchSnapshot()
    })
  })
})
