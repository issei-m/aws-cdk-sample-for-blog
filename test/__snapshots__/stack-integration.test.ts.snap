// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stack integration testing ApplicationStack synthesizes the way we expect 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "ApplicationFunction4738106C": {
      "DependsOn": [
        "ApplicationFunctionRole516B4A4D",
      ],
      "Properties": {
        "Architectures": [
          "arm64",
        ],
        "Code": {
          "ZipFile": "import os

def handler(event, context):
    return { 'bucket': os.environ['BUCKET_NAME'] }",
        },
        "Environment": {
          "Variables": {
            "BUCKET_NAME": {
              "Fn::ImportValue": "TestS3:ExportsOutputRefPublicUserUploadedFilesBucket4FF12C1B9072A75C",
            },
          },
        },
        "FunctionName": "test-application",
        "Handler": "index.handler",
        "MemorySize": 1024,
        "Role": {
          "Fn::GetAtt": [
            "ApplicationFunctionRole516B4A4D",
            "Arn",
          ],
        },
        "Runtime": "python3.12",
      },
      "Type": "AWS::Lambda::Function",
    },
    "ApplicationFunctionRole516B4A4D": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "RoleName": "test-application-function-role",
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`Stack integration testing S3Stack synthesizes the way we expect 1`] = `
{
  "Outputs": {
    "ExportsOutputRefPublicUserUploadedFilesBucket4FF12C1B9072A75C": {
      "Export": {
        "Name": "TestS3:ExportsOutputRefPublicUserUploadedFilesBucket4FF12C1B9072A75C",
      },
      "Value": {
        "Ref": "PublicUserUploadedFilesBucket4FF12C1B",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "PublicUserUploadedFilesBucket4FF12C1B": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "BucketName": "test-public-user-uploaded-files",
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
